### ISSUE TITLE
xetex: wrong position of node `current page`

### ISSUE THREAD

**Issue by muzimuzhi** [pgf-tikz/pgf#909]
**Brief outline of the bug**

From question https://tex.stackexchange.com/q/559138/79060.

Compiled with xetex, the position of node `current page` is wrongly calculated if `\hoffset` and/or `\voffset` are/is non-zero.

**Minimal working example (MWE)**

```latex
\documentclass{article}
\usepackage{tikz}

\usepackage[papersize={6cm, 8cm}]{geometry}
\hoffset=15pt
\voffset=15pt

\begin{document}
\begin{tikzpicture}[remember picture, overlay]
  \draw[line width=10pt, blue, double=blue!50, double distance=10pt] 
    (current page.north west) rectangle (current page.south east);
\end{tikzpicture}
\end{document}
```

Expected output: with pdftex or luatex
![image](https://user-images.githubusercontent.com/6376638/90532851-4cf47300-e1aa-11ea-8c47-e9b2e69dc644.png)

Actual output: with xetex
![image](https://user-images.githubusercontent.com/6376638/90532878-55e54480-e1aa-11ea-9509-9b47b4ef7818.png)

**Comment by muzimuzhi** [comment]
**Proposal**

```diff
diff --git a/tex/generic/pgf/systemlayer/pgfsys-xetex.def b/tex/generic/pgf/systemlayer/pgfsys-xetex.def
index e191b9da..9eb1d23d 100644
--- a/tex/generic/pgf/systemlayer/pgfsys-xetex.def
+++ b/tex/generic/pgf/systemlayer/pgfsys-xetex.def
@@ -62,7 +62,7 @@
   \pgf@marshal%
 }

-\def\pgf@sys@pdf@mark@pos@pgfpageorigin{\pgfqpoint{2\hoffset}{0pt}}
+\def\pgf@sys@pdf@mark@pos@pgfpageorigin{\pgfqpoint{\hoffset}{\voffset}}

 \ifx\paperheight\@undefined
 \else
```

The last modification to related line is done by https://github.com/pgf-tikz/pgf/commit/cdacb877e0dd2cefb2a5e1681ede483e4e5f16a8 in 2013, in order to fix #85. 

_Edit:_ I'm not familiar with either the system layer of `pgf` or `xetex` specific things, but an asymmetric use of `\hoffset` and `\voffset` is really wired and since the definitions of `\pgf@sys@pdf@mark@pos@pgfpageorigin` in every other `pgfsys-xxx.def` are symmetric ...
```tex
./pgfsys-common-pdf-via-dvi.def:110:  \def\pgf@sys@pdf@mark@pos@pgfpageorigin{\pgfpointorigin}
./pgfsys-dvips.def:158:\def\pgf@sys@pdf@mark@pos@pgfpageorigin{\pgfpointorigin}
./pgfsys-pdftex.def:266:\def\pgf@sys@pdf@mark@pos@pgfpageorigin{\pgfqpoint{\hoffset}{\voffset}}
./pgfsys-xetex.def:65:\def\pgf@sys@pdf@mark@pos@pgfpageorigin{\pgfqpoint{2\hoffset}{0pt}}
./pgfsys-dvipdfmx.def:250:  \def\pgf@sys@pdf@mark@pos@pgfpageorigin{\pgfpointorigin}
./pgfsys-dvisvgm.def:207:\def\pgf@sys@pdf@mark@pos@pgfpageorigin{\pgfpointorigin}
./pgfsys-luatex.def:269:\def\pgf@sys@pdf@mark@pos@pgfpageorigin{\pgfqpoint{\hoffset}{\voffset}}
```

**Comment by hmenke** [comment]
Exact duplicate of https://github.com/pgf-tikz/pgf/issues/294
Please search existing issues before opening a new one.

**Comment by muzimuzhi** [comment]
> Exact duplicate of #294

Not exactly the same. 

#294 reported a problem reproducible with every engine, while this issue reports a problem **specific to xetex**, which corresponds to a bug fix [1] of xetex introduced in texlive 2020. The above proposed change for `pgfsys-xetex.def` actually restores the buggy level of xetex to #294, but it will not and has no intent to fix #294.

[1] About this bug:
 - related bug report to xetex: https://sourceforge.net/p/xetex/bugs/56/ (it seems the bug is not fully resolved, I've reposted to Clerk Ma who provided the fix)
 - related commit to texlive: http://www.tug.org/svn/texlive?view=revision&revision=53855

Therefore the whole story is:
1. there was a long-standing xetex bug (not newer than xetex 0.999991)
1. pgf used some workaround to live with that bug
1. now the bug is fixed (since xetex 0.999992)
1. pgf need to take a corresponding change

A refined proposal would be
```diff
diff --git a/tex/generic/pgf/systemlayer/pgfsys-xetex.def b/tex/generic/pgf/systemlayer/pgfsys-xetex.def
index e191b9da..4cdaf9ab 100644
--- a/tex/generic/pgf/systemlayer/pgfsys-xetex.def
+++ b/tex/generic/pgf/systemlayer/pgfsys-xetex.def
@@ -62,7 +62,14 @@
   \pgf@marshal%
 }

-\def\pgf@sys@pdf@mark@pos@pgfpageorigin{\pgfqpoint{2\hoffset}{0pt}}
+% this test fails if xetex releases 1.0 in the future
+\ifnum\expandafter\pgfutil@gobble\XeTeXrevision>999991\relax
+  \def\pgf@sys@pdf@mark@pos@pgfpageorigin{\pgfqpoint{\hoffset}{\voffset}}
+\else
+  \def\pgf@sys@pdf@mark@pos@pgfpageorigin{\pgfqpoint{2\hoffset}{0pt}}
+\fi

 \ifx\paperheight\@undefined
 \else
```

Sorry I was obviously in a hurry yesterday. The #294 you mentioned let me learn more.

------
To be more specific, using the example given in https://github.com/pgf-tikz/pgf/issues/294#issuecomment-478804361,

```tex
\documentclass{article}
\usepackage[T1]{fontenc}
\usepackage{tikz}

\advance\hoffset by 4cm

\begin{document}
\tikz[overlay,remember picture]
  \draw[green,<->] (current page.south west) --  (current page.north east);%

abc
\end{document}
```

 - With texlive 2019 (for example using overleaf), each one of pdftex, xetex, and luatex gives the same shifted output.
 - With texlive 2020, output of pdftex and luatex are unchanged, but output of xetex now has the shifting amount doubled.
![image](https://user-images.githubusercontent.com/6376638/90723378-33a90f00-e2ef-11ea-84cd-a16c9e386ecc.png)![image](https://user-images.githubusercontent.com/6376638/90723238-f5135480-e2ee-11ea-866f-2045fc8877b7.png)

**Comment by hmenke** [comment]
Can you make a PR with that patch?
